crates/wasm-smith/src/lib.rs

```diff
@@ -2321,26 +2332,100 @@ impl Module {
 }
 
 /// This function generates a number between `min` and `max`, favoring values
-/// closer to `min`.
+/// between `min` and `max_inbounds`.
 ///
 /// The thinking behind this function is that it's used for things like offsets
 /// and minimum sizes which, when very large, can trivially make the wasm oom or
 /// abort with a trap. This isn't the most interesting thing to do so it tries
-/// to favor numbers closer to `min` if possible.
-fn gradually_grow(u: &mut Unstructured, min: u64, max: u64) -> Result<u64> {
-    // The idea behind how this works is that we use the input `u` to determin
-    // the ceiling, ideally between min/max, and then generate a number between
-    // min and our ceiling.
-    assert!(min <= max);
-    let mut cur_max = min.saturating_add(10);
-    loop {
-        if cur_max >= max {
-            break u.int_in_range(min..=max);
-        }
-        if u.int_in_range(0..=GRADUAL_FACTOR_OF_GROWTH)? == 0 {
-            break u.int_in_range(min..=cur_max);
-        }
-        cur_max += cur_max / 2;
+/// to favor numbers in the `min..max_inbounds` range to avoid immediate ooms.
+fn gradually_grow(u: &mut Unstructured, min: u64, max_inbounds: u64, max: u64) -> Result<u64> {
+    if min == max {
+        return Ok(min);
+    }
+    let min = min as f64;
+    let max = max as f64;
+    let max_inbounds = max_inbounds as f64;
+    let x = u.arbitrary::<u32>()?;
+    let x = f64::from(x);
+    let x = map_custom(
+        x,
+        f64::from(u32::MIN)..f64::from(u32::MAX),
+        min..max_inbounds,
+        min..max,
+    );
+    return Ok(x.round() as u64);
+
+    /// Map a value from within the input range to the output range(s).
+    ///
+    /// This will first map the input range into the `0..1` input range, and
+    /// then depending on the value it will either map it exponentially
+    /// (favoring small values) into the `output_inbounds` range or it will map
+    /// it into the `output` range.
+    fn map_custom(
+        value: f64,
+        input: Range<f64>,
+        output_inbounds: Range<f64>,
+        output: Range<f64>,
+    ) -> f64 {
+        assert!(!value.is_nan(), "{}", value);
+        assert!(value.is_finite(), "{}", value);
+        assert!(input.start < input.end, "{} < {}", input.start, input.end);
+        assert!(
+            output.start < output.end,
+            "{} < {}",
+            output.start,
+            output.end
+        );
+        assert!(value >= input.start, "{} >= {}", value, input.start);
+        assert!(value <= input.end, "{} <= {}", value, input.end);
```

Comment:
Maybe also assert that `output.start <= output_inbounds.start` and `output_inbounds.end <= output.end`?